File: backend/src/main/java/cn/etarch/mao/harness/tool/impl/WriteFileTool.java
package cn.etarch.mao.harness.tool.impl;

import cn.etarch.mao.harness.safety.PathSandbox;
import cn.etarch.mao.harness.tool.FileChangeDiffUtil;
import cn.etarch.mao.harness.tool.Tool;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.Map;

@Slf4j
@Component
public class WriteFileTool implements Tool {

    private final ObjectMapper objectMapper;
    private final PathSandbox pathSandbox;

    public WriteFileTool(ObjectMapper objectMapper, PathSandbox pathSandbox) {
        this.objectMapper = objectMapper;
        this.pathSandbox = pathSandbox;
    }

    @Override
    public String getName() {
        return "write_file";
    }

    @Override
    public String getDescription() {
        return "将内容写入文件（创建或覆盖）。如有需要会创建父目录。参数：path（必填）、content（必填）。";
    }

    @Override
    public Map<String, Object> getInputSchema() {
        Map<String, Object> schema = new HashMap<>();
        schema.put("type", "object");
        Map<String, Object> properties = new HashMap<>();
        properties.put("path", Map.of("type", "string", "description", "相对于工作区根目录的文件路径"));
        properties.put("content", Map.of("type", "string", "description", "要写入文件的内容"));
        schema.put("properties", properties);
        schema.put("required", new String[]{"path", "content"});
        return schema;
    }

    @Override
    public Map<String, Object> getOutputSchema() {
        Map<String, Object> schema = new HashMap<>();
        schema.put("type", "object");
        Map<String, Object> properties = new HashMap<>();
        properties.put("success", Map.of("type", "boolean"));
        properties.put("bytes_written", Map.of("type", "integer"));
        schema.put("properties", properties);
        return schema;
    }

    @Override
    public String execute(String arguments) {
        return execute(arguments, null);
    }

    @Override
    public String execute(String arguments, String workspace) {
        try {
            JsonNode args = objectMapper.readTree(arguments);
            String path = args.get("path").asText();
            String content = args.get("content").asText();

            Path filePath = pathSandbox.resolve(path, workspace);

            // Snapshot before write for change tracking
            boolean fileExisted = Files.exists(filePath);
            String beforeContent = "";
            if (fileExisted) {
                beforeContent = Files.readString(filePath);
            }

            Path parent = filePath.getParent();
            if (parent != null && !Files.exists(parent)) {
                Files.createDirectories(parent);
            }

            Files.writeString(filePath, content);

            // Compute file change stats
            int newLineCount = content.isEmpty() ? 0 : content.split("\n", -1).length;
            FileChangeDiffUtil.LineDelta lineDelta = fileExisted
                    ? FileChangeDiffUtil.computeLineDelta(beforeContent, content)
                    : new FileChangeDiffUtil.LineDelta(newLineCount, 0);
            int linesAdded = lineDelta.linesAdded();
            int linesDeleted = lineDelta.linesDeleted();
            String changeType = fileExisted ? "MODIFIED" : "CREATED";

            Map<String, Object> result = new HashMap<>();
            result.put("success", true);
            result.put("bytes_written", content.length());
            result.put("file_change", Map.of(
                    "path", path,
                    "type", changeType,
                    "total_lines", newLineCount,
                    "lines_added", linesAdded,
                    "lines_deleted", linesDeleted
            ));
            result.put(FileChangeDiffUtil.PRIVATE_DIFF_FIELD,
                    FileChangeDiffUtil.buildDiff(path, beforeContent, content));
            return objectMapper.writeValueAsString(result);
        } catch (IOException e) {
            log.error("WriteFileTool execution failed", e);
            try {
                return objectMapper.writeValueAsString(Map.of(
                        "success", false,
                        "bytes_written", 0,
                        "error", e.getMessage()
                ));
            } catch (Exception ex) {
                return "{\"success\":false,\"bytes_written\":0,\"error\":\"" + e.getMessage().replace("\"", "'") + "\"}";
            }
        }
    }
}

File: backend-ts/src/harness/tool/impl/write-file-tool.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import path from 'node:path';
import { BaseTool } from '../tool.js';
import { asText, parseObject, toJson } from '../json.js';
import { splitLines } from './read-file-tool.js';
import { FileChangeDiffUtil } from '../file-change-diff-util.js';
import type { PathSandbox } from '../../safety/path-sandbox.js';
import { harnessLog } from '../../log.js';
import { withFileLock } from './file-write-lock.js';
import { applyEol, detectEol, stripBom } from './file-eol.js';

export class WriteFileTool extends BaseTool {
  constructor(private readonly pathSandbox: PathSandbox) {
    super();
  }

  getName(): string { return 'write_file'; }
  getDescription(): string {
    return '将内容写入文件（创建或覆盖）。如有需要会创建父目录。参数：path（必填）、content（必填）。';
  }
  getInputSchema(): Record<string, unknown> {
    return {
      type: 'object',
      properties: {
        path: { type: 'string', description: '相对于工作区根目录的文件路径' },
        content: { type: 'string', description: '要写入文件的内容' },
      },
      required: ['path', 'content'],
    };
  }
  getOutputSchema(): Record<string, unknown> {
    return {
      type: 'object',
      properties: {
        success: { type: 'boolean' },
        bytes_written: { type: 'integer' },
      },
    };
  }

  protected async executeWithWorkspace(argumentsJson: string, workspace: string | null): Promise<string> {
    try {
      const args = parseObject(argumentsJson);
      if (!args) return toJson({ success: false, bytes_written: 0, error: '无效的JSON参数' });
      const filePathArg = asText(args.path);
      const content = asText(args.content);
      if (filePathArg == null || content == null) {
        return toJson({ success: false, bytes_written: 0, error: '缺少必填参数: path, content' });
      }
      const filePath = this.pathSandbox.resolve(filePathArg, workspace);
      // 与 edit_file 共用路径锁，避免并行写同一文件互相覆盖
      return await withFileLock(filePath, () => {
        const fileExisted = existsSync(filePath);
        const rawBefore = fileExisted ? readFileSync(filePath, 'utf8') : '';
        const beforeContent = stripBom(rawBefore);
        const toWrite = fileExisted ? applyEol(content, detectEol(rawBefore)) : content;
        const parent = path.dirname(filePath);
        if (parent && !existsSync(parent)) mkdirSync(parent, { recursive: true });
        writeFileSync(filePath, toWrite);
        const newLineCount = splitLines(content).length;
        const lineDelta = fileExisted
          ? FileChangeDiffUtil.computeLineDelta(beforeContent, content)
          : { linesAdded: newLineCount, linesDeleted: 0 };
        return toJson({
          success: true,
          bytes_written: Buffer.byteLength(toWrite, 'utf8'),
          file_change: {
            path: filePathArg,
            type: fileExisted ? 'MODIFIED' : 'CREATED',
            total_lines: newLineCount,
            lines_added: lineDelta.linesAdded,
            lines_deleted: lineDelta.linesDeleted,
          },
          [FileChangeDiffUtil.PRIVATE_DIFF_FIELD]: FileChangeDiffUtil.buildDiff(filePathArg, beforeContent, content),
        });
      });
    } catch (e) {
      harnessLog('error', 'WriteFileTool execution failed', e);
      return toJson({ success: false, bytes_written: 0, error: (e as Error).message });
    }
  }
}
